### ISSUE TITLE
Improve cache handling

### ISSUE THREAD

**Issue by herm** [python-ivi/python-ivi#43]
Currently a lot of functions are almost identical, because they mainly consist of cache handling code. Using decorators most of this code duplication could be avoided. I implemented a demonstration for this.

This very long code
```
    def _get_trigger_delay(self):
        if not self._driver_operation_simulate and not self._get_cache_valid():
            value = float(self._ask("trigger:delay?"))
            self._trigger_delay = value
            self._set_cache_valid()
        return self._trigger_delay
    
    def _set_trigger_delay(self, value):
        value = float(value)
        if not self._driver_operation_simulate:
            self._write('trigger:delay %g' % value)
        self._trigger_delay = value
        self._set_cache_valid()
``` 
becomes
```
    @cacheable
    def _get_trigger_delay(self):
        return value = float(self._ask("trigger:delay?"))

    @cacheable    
    def _set_trigger_delay(self, value):
        self._write('trigger:delay %g' % value)
```
Which is much easier to read an less error-prone. At the moment this is just a demonstration, but if there is interest I can implement a real solution which is compatible with the current caching mechanism. 
Code:  https://gist.github.com/herm/9e9d02218f5a1dddd8d9

**Comment by alexforencich** [comment]
Yeah, something like this makes a good deal of sense, if only to remove the stack introspection.  Don't forget about indexed properties, though.  And there needs to be a good way of invalidating the cached values of other attributes - e.g. invalidate scope channel range when probe gain is changed.  I think adding some extensible support for coercion (round to a particular resolution and/or pick nearest/next largest/next smallest from a list) would be a good idea as well, and possibly combine it with caching.  The other thing that I need to implement is proper overriding for sets and mappings.  Right now that does not work.  Also, there needs to be a good way to query what the allowed values are for various settings, especially those which only allow a specific set.
